servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java

```diff
@@ -240,21 +230,26 @@ private static void updateResponseContentLength(final int contentLength, final H
         });
     }
 
-    static StreamingHttpResponse addResponseTransferEncodingIfNecessary(final StreamingHttpResponse response,
-                                                                        final HttpRequestMethod requestMethod) {
-        if (canAddResponseTransferEncoding(response, requestMethod)) {
+    static void addResponseTransferEncodingIfNecessary(final StreamingHttpResponse response,
+                                                       final HttpRequestMethod requestMethod) {
+        if (canAddResponseTransferEncodingProtocol(response.status().code(), requestMethod) &&
+                canAddTransferEncodingChunked(response)) {
             response.headers().add(TRANSFER_ENCODING, CHUNKED);
         }
-        return response;
     }
 
     static void addRequestTransferEncodingIfNecessary(final StreamingHttpRequest request) {
-        if (canAddRequestTransferEncoding(request)) {
+        if (clientMaySendPayloadBodyFor(request.method()) && canAddTransferEncodingChunked(request)) {
             request.headers().add(TRANSFER_ENCODING, CHUNKED);
         }
     }
 
-    private static boolean hasContentHeaders(final HttpHeaders headers) {
+    private static boolean canAddTransferEncodingChunked(HttpMetaData metaData) {
+        final HttpHeaders headers = metaData.headers();
+        return !isTransferEncodingChunked(headers) && (!headers.contains(CONTENT_LENGTH) || mayHaveTrailers(metaData));
```

Comment:
should we also take into account the protocol version (if available, similar to `canAddContentLength`)? for h2 we never need to add transfer encoding.